src/Raven.Server/Documents/ShardedHandlers/ShardedBatchHandler.cs

```diff
@@ -0,0 +1,355 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Documents.Changes;
+using Raven.Client.Documents.Commands.Batches;
+using Raven.Client.Documents.Session;
+using Raven.Client.Http;
+using Raven.Client.Json;
+using Raven.Server.Documents.Handlers;
+using Raven.Server.Documents.ShardedHandlers.ShardedCommands;
+using Raven.Server.Documents.Sharding;
+using Raven.Server.Routing;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.TrafficWatch;
+using Raven.Server.Web;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Documents.ShardedHandlers
+{
+    public class ShardedBatchHandler : ShardedRequestHandler
+    {
+        [RavenShardedAction("/databases/*/bulk_docs", "POST")]
+        public async Task BulkDocs()
+        {
+            var commandBuilder = new ShardedBatchCommandBuilder(this, ShardedContext.DatabaseName, ShardedContext.IdentitySeparator);
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var contentType = HttpContext.Request.ContentType;
+                if (contentType == null ||
+                    contentType.StartsWith("application/json", StringComparison.OrdinalIgnoreCase))
+                {
+                    await commandBuilder.BuildCommandsAsync(context, RequestBodyStream());
+                }
+                else if (contentType.StartsWith("multipart/mixed", StringComparison.OrdinalIgnoreCase) ||
+                         contentType.StartsWith("multipart/form-data", StringComparison.OrdinalIgnoreCase))
+                {
+                    await commandBuilder.ParseMultipart(context, RequestBodyStream(), HttpContext.Request.ContentType);
+                }
+                else
+                    BatchHandler.ThrowNotSupportedType(contentType);
+
+                if (TrafficWatchManager.HasRegisteredClients)
+                {
+                    var log = BatchHandler.BatchTrafficWatch(commandBuilder.Commands);
+                    // add sb to httpContext
+                    AddStringToHttpContext(log, TrafficWatchChangeType.BulkDocs); // TODO sharding: maybe ShardedBulkDocs type?
+                }
+
+                using (var batch = await commandBuilder.GetCommand(context))
+                {
+                    if (batch.IsClusterTransaction)
+                    {
+                        throw new NotSupportedException("Cluster transactions are currently not supported in a sharded database.");
+                    }
+
+                    var shardedBatchCommands = new Dictionary<int, SingleNodeShardedBatchCommand>();
```

Comment:
I wonder if we can / should cache these.

Reply:
will add a todo